src/loki/search.cc

```diff
@@ -120,21 +120,42 @@ struct candidate_t {
     return sq_distance < c.sq_distance;
   }
 
-  PathLocation::SideOfStreet
-  get_side(const PointLL& original, double sq_distance, double sq_tolerance) const {
-    // its so close to the edge that its basically on the edge
-    if (sq_distance < sq_tolerance) {
+  PathLocation::SideOfStreet get_side(const PointLL& original,
+                                      double sq_distance,
+                                      double sq_tolerance,
+                                      double sq_max_distance) const {
+    // point is so close to the edge that its basically on the edge,
+    // or its too far from the edge that we shouldn't use it to determine side of street
+    if (sq_distance < sq_tolerance || sq_distance > sq_max_distance) {
       return PathLocation::SideOfStreet::NONE;
     }
 
-    // get the side TODO: this can technically fail for longer segments..
-    // to fix it we simply compute the plane formed by the triangle
-    // through the center of the earth and the two shape points and test
-    // whether the original point is above or below the plane (depending on winding)
+    // get the angle of the tangent line to the shape at this point
     auto& shape = edge_info->shape();
-    LineSegment2<PointLL> segment(shape[index], shape[index + 1]);
-    return (segment.IsLeft(original) > 0) == edge->forward() ? PathLocation::SideOfStreet::LEFT
-                                                             : PathLocation::SideOfStreet::RIGHT;
+    auto tang_angle =
+        tangent_angle(index, point, shape, GetOffsetForHeading(edge->classification(), edge->use()),
+                      edge->forward());
+
+    // get the absolute angle between the snap point and the provided point
+    auto angle_to_point = point.Heading(original);
+
+    // add 360 degrees if angle becomes negative
+    auto angle_diff = angle_to_point - tang_angle;
+    if (angle_diff < 0) {
+      angle_diff += 360.f;
+    }
+
+    // 10 degrees on either side is considered to be straight ahead
+    auto angle_tolerance = 10.f;
+
+    // check which side the point falls in
+    if (angle_diff > angle_tolerance && angle_diff < (180.f - angle_tolerance)) {
```

Comment:
Can we use the raw angles here, similar to how its done in [odin](https://github.com/valhalla/valhalla/blob/master/src/odin/maneuversbuilder.cc#L2237) & [baldr](https://github.com/valhalla/valhalla/blob/master/src/baldr/turn.cc#L20)? IMO its a bit easier to visualize. You can leave it as-is if you think we will be tweaking `angle_tolerance` often (which will cause recalculation of the hard-coded angles).

Reply:
Rest of the calc looks good 👍

Reply:
I don't think the angle_tolerance will change often but I'm hesitant to hardcode the value without giving it a meaningful name. I added a comment block with the exact values assuming angle_tolerance=10 as well as an ascii diagram, let me know if that helps visualize it.

Reply:
The diagram def helps! 👍